tracer/src/Datadog.Trace/Iast/ReturnedHeadersAnalyzer.cs

```diff
@@ -39,6 +41,140 @@ internal static void Analyze(IHeaderDictionary responseHeaders, IntegrationId in
+    // Tainted range is considered sensitive value
+
+#if NETFRAMEWORK
+    private static void AnalyzeHeaderInjectionVulnerability(NameValueCollection responseHeaders, IntegrationId integrationId)
+#else
+    private static void AnalyzeHeaderInjectionVulnerability(IHeaderDictionary responseHeaders, IntegrationId integrationId)
+#endif
+    {
+        try
+        {
+            IastModule.OnExecutedSinkTelemetry(IastInstrumentedSinks.HeaderInjection);
+
+            var tracer = Tracer.Instance;
+            var currentSpan = (tracer.ActiveScope as Scope)?.Span;
+            var traceContext = currentSpan?.Context?.TraceContext;
+
+#if NETFRAMEWORK
+            foreach (var headerKey in responseHeaders.AllKeys)
+#else
+            foreach (var headerKey in responseHeaders.Keys)
+#endif
+            {
+                if (string.IsNullOrWhiteSpace(headerKey))
+                {
+                    continue;
+                }
+
+                bool isHeaderInjectionException = false;
+                foreach (var excludeType in headerInjectionExceptions)
+                {
+                    if (excludeType.Equals((string)headerKey, StringComparison.OrdinalIgnoreCase))
+                    {
+                        isHeaderInjectionException = true;
+                        break;
+                    }
+                }
+
+                if (isHeaderInjectionException)
+                {
+                    continue;
+                }
+
+                string headerValue = responseHeaders[headerKey];
+
+                if (string.IsNullOrWhiteSpace(headerValue))
+                {
+                    continue;
+                }
+
+                var taintedValue = traceContext?.IastRequestContext?.GetTainted(headerValue);
+
+                if (taintedValue is null || taintedValue.Ranges.Length == 0)
+                {
+                    continue;
+                }
+
+                if (headerKey.Equals("set-cookie", StringComparison.OrdinalIgnoreCase))
+                {
+                    if (OnlyCookieValueSources(taintedValue))
+                    {
+                        continue;
+                    }
+                }
+
+                if (headerKey.StartsWith("access-control-allow-", StringComparison.OrdinalIgnoreCase))
+                {
+                    if (ComesFromOriginHeader(taintedValue))
+                    {
+                        continue;
+                    }
+                }
+
+                if (IsPropagationHeader(headerKey, taintedValue))
+                {
+                    continue;
+                }
+
+                IastModule.OnHeaderInjection(integrationId, headerKey, headerValue);
+            }
+        }
+        catch (Exception error)
+        {
+            Log.Error(error, $"Error in {nameof(ReturnedHeadersAnalyzer)}.{nameof(AnalyzeHeaderInjectionVulnerability)}");
+        }
+    }
+
+    private static bool OnlyCookieValueSources(TaintedObject taintedValue)
```

Comment:
nit: `OnlyContainsCookieValueSources` - I thought it was filtering initially 🙂

Reply:
Done!